File: algorithms/3_hard/S/01406_stone_game_iii/dp.c
static inline int max_int(const int a, const int b) {
    return a > b ? a : b;
}

char* stoneGameIII(const int* stone_value, const int len) {
    if (len < 0 || (len > 0 && stone_value == NULL)) {
        fprintf(stderr, "invalid input");
        exit(1);
    }
    int next1 = 0;
    int next2 = 0;
    int next3 = 0;

    for (int i = len - 1; i >= 0; i -= 1) {
        int curr = stone_value[i] - next1;
        if (i + 1 < len) {
            curr = max_int(curr, stone_value[i] + stone_value[i + 1] - next2);
        }
        if (i + 2 < len) {
            curr = max_int(curr, stone_value[i] + stone_value[i + 1] + stone_value[i + 2] - next3);
        }
        next3 = next2;
        next2 = next1;
        next1 = curr;
    }

    if (next1 < 0) {
        return strdup("Bob");
    } else if (next1 == 0) {
        return strdup("Tie");
    } else {
        return strdup("Alice");
    }
}

File: algorithms/3_hard/S/01406_stone_game_iii/dp.rs
impl Solution {
    pub fn stone_game_iii(stone_value: Vec<i32>) -> String {
        let len = stone_value.len();
        let (mut next1, mut next2, mut next3) = (0_i32, 0_i32, 0_i32);

        for i in (0..len).rev() {
            let mut curr = stone_value[i] - next1;
            if i + 1 < len {
                curr = curr.max(stone_value[i] + stone_value[i + 1] - next2);
            }
            if i + 2 < len {
                curr = curr.max(stone_value[i] + stone_value[i + 1] + stone_value[i + 2] - next3);
            }
            (next1, next2, next3) = (curr, next1, next2);
        }

        match next1.cmp(&0) {
            std::cmp::Ordering::Less => "Bob".to_string(),
            std::cmp::Ordering::Equal => "Tie".to_string(),
            std::cmp::Ordering::Greater => "Alice".to_string(),
        }
    }
}
